File: src/cli/commands/rm.ts
/**
 * Remove command implementation
 * 
 * Removes a host entry by hostname from ~/.hosts
 */

import { readHostsFile, writeHostsFile } from "../../core/file-io";
import type { HostsFile, Group } from "../../domain/types";
import { ExitCode } from "../exit-codes";

/**
 * Execute the rm command
 * 
 * @param hostname - Hostname to remove
 * @returns Exit code (0 = success, 1 = not found, 2 = I/O error)
 */
export async function rmCommand(hostname: string): Promise<number> {
  try {
    // Read ~/.hosts
    const hostsFile = await readHostsFile("~/.hosts");

    // Find and remove the entry
    let found = false;
    const updatedHostsFile = removeEntryByHostname(hostsFile, hostname, (wasFound) => {
      found = wasFound;
    });

    // If not found, return exit code 1
    if (!found) {
      console.error(`Error: Hostname '${hostname}' not found`);
      return ExitCode.VALIDATION;
    }

    // Write back to ~/.hosts
    await writeHostsFile("~/.hosts", updatedHostsFile);

    console.log(`✓ Removed '${hostname}'`);
    return ExitCode.SUCCESS;

  } catch (err: any) {
    // Handle I/O errors
    console.error(`Error: ${err.message}`);
    return ExitCode.IO_ERROR;
  }
}

/**
 * Recursively remove an entry by hostname from a HostsFile
 * 
 * @param hostsFile - The hosts file structure
 * @param hostname - Hostname to remove
 * @param onFound - Callback to signal if entry was found
 * @returns Updated HostsFile with entry removed
 */
function removeEntryByHostname(
  hostsFile: HostsFile,
  hostname: string,
  onFound: (found: boolean) => void
): HostsFile {
  let found = false;

  const updatedGroups = hostsFile.groups.map(group => 
    removeFromGroup(group, hostname, (wasFound) => {
      if (wasFound) found = true;
    })
  );

  onFound(found);

  return {
    ...hostsFile,
    groups: updatedGroups,
  };
}

/**
 * Recursively remove an entry from a group and its nested groups
 * 
 * @param group - The group to process
 * @param hostname - Hostname to remove
 * @param onFound - Callback to signal if entry was found
 * @returns Updated Group with entry removed
 */
function removeFromGroup(
  group: Group,
  hostname: string,
  onFound: (found: boolean) => void
): Group {
  // Filter entries in this group
  const filteredEntries = group.entries.filter(entry => {
    if (entry.hostname === hostname) {
      onFound(true);
      return false;
    }
    return true;
  });

  // Recursively process nested groups
  const updatedNestedGroups = group.groups.map(nestedGroup =>
    removeFromGroup(nestedGroup, hostname, onFound)
  );

  return {
    ...group,
    entries: filteredEntries,
    groups: updatedNestedGroups,
  };
}

File: go/internal/cmd/rm.go
package cmd

import (
	"fmt"

	"github.com/hungthai1401/hostie/go/internal/core/fileio"
	"github.com/hungthai1401/hostie/go/internal/domain"
	"github.com/spf13/cobra"
)

func newRmCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "rm <hostname>",
		Short: "Remove a hosts entry",
		Long:  "Remove an entry from the hosts file by hostname",
		Args:  cobra.ExactArgs(1),
		RunE:  runRm,
	}

	return cmd
}

func runRm(cmd *cobra.Command, args []string) error {
	hostname := args[0]

	// Read existing hosts file
	hostsFile, err := fileio.ReadHostsFile(hostsFilePath)
	if err != nil {
		return fmt.Errorf("failed to read hosts file: %w", err)
	}

	// Remove entry by hostname
	found := removeEntryByHostname(&hostsFile, hostname)
	if !found {
		return fmt.Errorf("entry not found: %s", hostname)
	}

	// Write back to file
	if err := fileio.WriteHostsFile(hostsFilePath, hostsFile); err != nil {
		return fmt.Errorf("failed to write hosts file: %w", err)
	}

	fmt.Printf("✓ Removed %s\n", hostname)
	return nil
}

// removeEntryByHostname removes an entry from the hosts file by hostname
// Returns true if the entry was found and removed, false otherwise
func removeEntryByHostname(hf *domain.HostsFile, hostname string) bool {
	for i := range hf.Groups {
		if removeEntryFromGroup(&hf.Groups[i], hostname) {
			return true
		}
	}
	return false
}

// removeEntryFromGroup recursively removes an entry from a group and its subgroups
func removeEntryFromGroup(g *domain.Group, hostname string) bool {
	// Check entries in this group
	for i, entry := range g.Entries {
		if entry.Hostname == hostname {
			// Remove entry by slicing
			g.Entries = append(g.Entries[:i], g.Entries[i+1:]...)
			return true
		}
	}

	// Check subgroups
	for i := range g.Groups {
		if removeEntryFromGroup(&g.Groups[i], hostname) {
			return true
		}
	}

	return false
}
